libp2p/protocols/identify.nim

```diff
@@ -112,10 +112,15 @@ proc newIdentify*(peerInfo: PeerInfo): Identify =
 
 method init*(p: Identify) =
   proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
-    trace "handling identify request"
-    var pb = encodeMsg(p.peerInfo, await conn.getObservedAddrs())
-    await conn.writeLp(pb.buffer)
-    # await conn.close() #TODO: investigate why this breaks
+    try:
+      try:
```

Comment:
I don't understand the double try here, can't we just try->except->finally ?

Reply:
double try is a `defer`, if close throws for some reason, we're still going to catch it by the outter `try/except`.

Reply:
If `close` can throw, this code is not safe, because `finally` may be executed during stack unwinding of another exception. You'll have a situation where you are raising a new exception while another exception is still looking for a handler. Nim doesn't yet define what the behavior of this situation should be. In other languages such as C++, the program will be terminated at this point.

Reply:
In general, this is a complicated problem to deal with. The usual advice is that operations such as `close` should not raise exceptions, but there are some other possible strategies as well:

1) Check if finally is being executed during stack unwinding or due to normal exiting from the function:

```nim
if getCurrentException() == nil: ...
```

2) Wrap the `close` in a try/except. If this was a normal exit from the function, you may decide to re-raise the exception. Otherwise, debug log it and discard it.

Reply:
in general, `close` doesn't throw, so this (albeit ugly) syntax, should emulate what 

``` nim
try:
  defer: 
    await conn.close()
except:
   ...
```

would usually do. 

So I believe as it stands, this is safe?

Reply:
If `close` doesn't throw, you are safe. But your comment above was saying "if close throws for some reason", hence my reply.